### PR TITLE
MySQL: wipe the database between tests

### PR DESCRIPTION
**Pull request by MaxGabriel**
Initial implementation of wiping the database for the MySQL backend (I guess I'll end up doing 1 PR per-backend?). Based on testing this works fine, though [the logs](https://gist.github.com/MaxGabriel/87d5ca66a0959df1e761) are a little wonky; sometimes it seems to be missing the last few SQL statements from a test in the logs. For example, these logs are missing the `TRUNCATE TABLE user` and `SET foreign_key_checks = 1` queries:

```
11/Jan/2015:21:14:41 -0800 [Debug#SQL] "SHOW TABLES;" [] @(persistent-2.1.1.3:Database.Persist.Sql.Raw ./Database/Persist/Sql/Raw.hs:42:26)
11/Jan/2015:21:14:41 -0800 [Debug#SQL] "SET foreign_key_checks = 0;" [] @(persistent-2.1.1.3:Database.Persist.Sql.Raw ./Database/Persist/Sql/Raw.hs:64:18)
11/Jan/2015:21:14:41 -0800 [Debug#SQL] "TRUNCATE TABLE email" [] @(persistent-2.1.1.3:Database.Persist.Sql.Raw ./Database/Persist/Sql/Raw.hs:64:18)
Tables are ["email","user"]
11/Jan/2015:21:14:41 -0800 [Debug#SQL] "SHOW TABLES;" [] @(persistent-2.1.1.3:Database.Persist.Sql.Raw ./Database/Persist/Sql/Raw.hs:42:26)
```

It seems to work though, so maybe the logging is just dropping some messages or something.

Probably should wait until I get PRs up for all the backends before merging any, since they'll have some common code.

### CODE REVIEW

**Comment by creichert** [comment]
Looks great. A few initial thoughts:

- Might be worth adding some comments on the added functions about what is exactly truncated and removed. It might not be clear to those who know less about Sql.

- Should this be configurable or "opt-in"? Sometimes I do want to keep my test db between iterations for various reasons.

**Comment by MaxGabriel** [comment]
@creichert Good idea about the comments; I'll add those just like the rest of the scaffolding has.

re: opting in, since this code is added to the scaffolding it's possible to change it as necessary, or remove it entirely. For example, if you had a static `countries` table that you didn't want truncated, you could remove that from the list after the call to `getTables`. 

As far as keeping the test DB for some tests, maybe one could add a parameter to `withApp` for whether or not to wipe the DB for that set of tests? I haven't thought about this much.

**Comment by gregwebs** [comment]
not wiping the database leads to dependencies between tests so it usually isn't considered good practice. If you want to not wipe the database sometimes you can merge the separate test cases into one with multiple assertions. So I wouldn't worry about it for this pull request.

**Comment by snoyberg** [comment]
Looks good to me. @MaxGabriel let me know when the docs are in place and I should merge.

**Comment by MaxGabriel** [comment]
Ok, I did just a touch of refactoring and added comments. @creichert Do you have input on if these comments explain things well enough?

**Comment by creichert** [comment]
@MaxGabriel Awesome, looks great to me!

File: test/TestImport.hs [hunk 1]
```diff
@@ -24,4 +27,28 @@ withApp = before $ do
...
+        tables <- getTables
+        let queries = map ("TRUNCATE TABLE " ++ ) tables
+
+        -- In MySQL, a table cannot be truncated if another table references it via foreign key.
+        -- Since we're wiping both the parent and child tables, though, it's safe
+        -- to temporarily disable this check.
+        rawExecute "SET foreign_key_checks = 0;" []
+        forM_ queries (\q -> rawExecute q [])
+        rawExecute "SET foreign_key_checks = 1;" []
+    return ()
```
Comment (snoyberg) [inline test/TestImport.hs]: Is this line actually necessary?

File: test/TestImport.hs [hunk 2]
```diff
@@ -24,4 +27,28 @@ withApp = before $ do
...
+
+getTables :: MonadIO m => ReaderT SqlBackend m [Text]
+getTables = do
+    tables <- rawSql "SHOW TABLES;" []
```
Comment (snoyberg) [inline test/TestImport.hs]: I'm slightly concerned about this, since it allows the tests to wipe out tables they never populated (i.e., tables that Persistent does not know about). On the other hand, that may be exactly the right behavior for the test suite, and someone who has a problem with that can override the behavior anyway.

File: test/TestImport.hs (hunk 1 above)
Comment (MaxGabriel) [inline test/TestImport.hs]: Oh! Its not; I'll remove it.

File: test/TestImport.hs (hunk 2 above)
Comment (MaxGabriel) [inline test/TestImport.hs]: So I'd be super concerned in a production setting, but for a test environment I think its unlikely we'd do real damage.

The major case I can think of where you want a table not wiped is when you use a migration tool that maintains a table like `schema_migrations`. For our rails codebase at work, we have Jenkins run the migrations as a (partial) way of checking if they'll fail before we run them against production. For that to work, I'm pretty sure it needs `schema_migrations` to not be wiped. (this doesn't apply to a clean-slate environment like Travis CI). I'll look into this, but I suspect there isn't a general solution besides developers blacklisting tables themselves.

One disadvantage to not wiping all the tables is that it might leave the database in an inconsistent state with respect to referential integrity. If you have table A that persistent knows about, table B that it doesn't, and table B has a foreign key to table A, then wiping table A leaves the database inconsistent. For a database like Postgres, it'll raise an error if table B isn't in the list of tables passed to `TRUNCATE`; for MySQL I'd have to check if there's a decent solution. This seems edge-casey though.

Anyway I need to start work, will come back to this later tonight.

Comment (MaxGabriel) [inline test/TestImport.hs]: It looks like tools like DatabaseCleaner for Ruby just [special case `schema_migrations`](https://github.com/DatabaseCleaner/database_cleaner/blob/a1ac42cf41ea35f2bd88d507e7d2eb323ff1b679/lib/database_cleaner/active_record/deletion.rb#L58). That wouldn't be a huge deal for someone using a schema migration tool to blacklist, though.

I can look into getting the table names from the Persistent DSL if you'd like.

Comment (snoyberg) [inline test/TestImport.hs]: I'm OK leaving it as-is for now. I just realized one last issue: table names need to be escaped. I believe you can use [connEscapeName](http://www.stackage.org/haddock/lts-1.0/persistent-2.1.1.3/Database-Persist-Sql.html#v:connEscapeName) for that. Once that's done, I think this can be merged in. Thanks!

**Comment by MaxGabriel** [comment]
Ok, I think this is ready to go in. Postgres is almost ready, it should use the `connEscapeName` function as well. I'm not sure if Mongo has been reviewed, but it should be ready otherwise.

Still working on SQLite (running into this weird `SQLite3 returned ErrorConstraint while attempting to perform step` exception).

**Comment by snoyberg** [comment]
Thanks for the code and discussion guys!

**Comment by MaxGabriel** [comment]
Thanks for reviewing the code!

**Comment by MaxGabriel** [comment]
Hm, I think my SQLite truncation is failing because disabling foreign keys (`PRAGMA foreign_keys = OFF`) is a noop in a transaction. Anyone have a good way of running SQL with persist without one? Its looking like most everything gets down to `runSqlConn` which starts a transaction.

**Comment by snoyberg** [comment]
For SQLite, what about just deleting the database file itself each time?

**Comment by MaxGabriel** [comment]
Well, this works for disabling foreign keys for a database wipe in SQLite, just trying to clean it up + enable logging for that bit:

```haskell
cleanup :: App -> IO ()
cleanup app = do
    let settings = appSettings app   
    sqliteConn <- rawConnection (sqlDatabase $ appDatabaseConf settings)    
    disableForeignKeys sqliteConn

    pool <- runNoLoggingT $ createSqlPool (wrapConnection sqliteConn) 1

    flip runSqlPersistMPool pool $ do
        tables <- getTables
        sqlBackend <- ask
        let queries = map (\t -> "DELETE FROM " ++ (connEscapeName sqlBackend $ DBName t)) tables
        forM_ queries (\q -> rawExecute q [])
    return ()

rawConnection :: Text -> IO Sqlite.Connection
rawConnection t = Sqlite.open t

disableForeignKeys :: Sqlite.Connection -> IO ()
disableForeignKeys conn = Sqlite.prepare conn "PRAGMA foreign_keys = OFF;" >>= void . Sqlite.step

getTables :: MonadIO m => ReaderT SqlBackend m [Text]
getTables = do
    tables <- rawSql "SELECT name FROM sqlite_master WHERE type = 'table';" []
    return (fmap unSingle tables)
```

Deleting the DB file could be an option. One downside is that you then need to run a bunch of migrations after each test. Possible downside there is that if you're using an external tool to manage migrations (besides Yesod), then you have to set that up to run before each test as well. Also it just seems weird to me. 

Unless you made a copy of the SQLite database to begin with, then deleted it after each test and used the copy to make a new one (?).

Still, I think the code above isn't too bad.
